executor.js

```diff
@@ -98,55 +98,42 @@ async function handleInput(line, codeCaller) {
   };
 }
 
-// Our overall loop looks like this: read a line of input from stdin, spin
-// off a python worker to handle it, and write the results back to stdout.
-const rl = readline.createInterface({
-  input: process.stdin,
-  output: process.stdout,
-  terminal: false,
-});
-
 let questionTimeoutMilliseconds = Number.parseInt(process.env.QUESTION_TIMEOUT_MILLISECONDS);
 if (Number.isNaN(questionTimeoutMilliseconds)) {
   questionTimeoutMilliseconds = 10000;
 }
 
-let codeCaller = new CodeCallerNative({
-  dropPrivileges: true,
-  questionTimeoutMilliseconds,
-  errorLogger: console.error,
-});
-codeCaller.ensureChild();
+(async () => {
+  let codeCaller = new CodeCallerNative({
+    dropPrivileges: true,
+    questionTimeoutMilliseconds,
+    errorLogger: console.error,
+  });
+  await codeCaller.ensureChild();
 
-// Safety check: if we receive more input while handling another request,
-// discard it.
-let processingRequest = false;
-rl.on('line', (line) => {
-  if (processingRequest) {
-    // Someone else messed up - fail fast.
-    process.exit(1);
-  }
+  // Our overall loop looks like this: read a line of input from stdin, spin
+  // off a python worker to handle it, and write the results back to stdout.
+  const rl = readline.createInterface({
+    input: process.stdin,
+    output: process.stdout,
+    terminal: false,
+  });
 
-  processingRequest = true;
-  handleInput(line, codeCaller)
-    .then((results) => {
-      const { needsFullRestart, ...rest } = results;
-      if (needsFullRestart) {
-        codeCaller.done();
-        codeCaller = new CodeCallerNative();
-        codeCaller.ensureChild();
-      }
-      console.log(JSON.stringify(rest));
-      processingRequest = false;
-    })
-    .catch((err) => {
-      console.error(err);
-      processingRequest = false;
-    });
-});
+  // Once the readline interface closes, we can't get any more input; die
+  // immediately to allow our container to be removed.
+  rl.on('close', () => process.exit(0));
 
-rl.on('close', () => {
-  // We can't get any more input; die immediately to allow our container
-  // to be removed.
-  process.exit(0);
+  for await (const line of rl) {
+    const results = await handleInput(line, codeCaller);
+    const { needsFullRestart, ...rest } = results;
```

Comment:
Did you deliberately move this above the `needsFullRestart` handler? Any chance of a race condition here?

Reply:
This was deliberate; writing data back to stdout helps ensure that the parent process gets useful information instead of a potential timeout due to `ensureChild()` taking a long time.

I don't believe there's a chance for a race condition; is there a specific scenario you're asking about?

Reply:
I was thinking that after we write to stdout this signals the caller that the call has completed, so they might write a new line to our stdin, which might then trigger us to do something while the full restart is only partly completed? I have not thought this through with any care so this was just a general question to make sure you had thought about it 😄

Reply:
The rewrite to use an async iterator should prevent against that; we'll only repeat the loop once the `ensureChild()` call completes. Async iterators are really nice in that they have normal loop semantics, i.e. it'll never run the body of your loop multiple times concurrently 😛

Reply:
Excellent point, thanks for clarifying!